src/odb/include/odb/db.h

```diff
@@ -7642,22 +7659,110 @@ class dbPowerSwitch : public dbObject
 class dbScanChain : public dbObject
 {
  public:
+  dbSet<dbScanPartition> getScanPartitions() const;
+
+  dbSet<dbScanList> getScanLists() const;
+
+  // User Code Begin dbScanChain
+  const std::string& getName() const;
+
+  void setName(std::string_view name);
+
+  void setScanIn(dbBTerm* scan_in);
+  void setScanIn(dbITerm* scan_in);
+  std::variant<dbBTerm*, dbITerm*> getScanIn() const;
+
+  void setScanOut(dbBTerm* scan_out);
+  void setScanOut(dbITerm* scan_out);
+  std::variant<dbBTerm*, dbITerm*> getScanOut() const;
+
+  void setScanEnable(dbBTerm* scan_enable);
+  void setScanEnable(dbITerm* scan_enable);
+  std::variant<dbBTerm*, dbITerm*> getScanEnable() const;
+
+  void setTestMode(dbBTerm* test_mode);
+  void setTestMode(dbITerm* test_mode);
+  std::variant<dbBTerm*, dbITerm*> getTestMode() const;
+
+  void setTestModeName(const std::string& test_mode_name);
+  const std::string& getTestModeName() const;
+
+  static dbScanChain* create(dbDft* dft);
+  // User Code End dbScanChain
 };
 
 class dbScanInst : public dbObject
 {
  public:
-  enum class SCAN_INST_TYPE
+  struct AccessPins
   {
-    OneBit,
-    ShiftRegister,
-    BlackBox
+    std::variant<dbBTerm*, dbITerm*> scan_in;
+    std::variant<dbBTerm*, dbITerm*> scan_out;
   };
+  enum class ClockEdge
+  {
+    Rising,
+    Falling
+  };
+
+  // User Code Begin dbScanInst
+  void setScanClock(std::string_view scan_clock);
+  const std::string& getScanClock() const;
+
+  void setClockEdge(ClockEdge clock_edge);
+  ClockEdge getClockEdge() const;
+
+  // The number of bits that are in this scan inst from the scan in to the scan
+  // out. For simple flops this is just 1.
+  void setBits(uint bits);
+  uint getBits() const;
+
+  void setScanEnable(dbBTerm* scan_enable);
```

Comment:
What does it mean for the enable of a scan inst to be a bterm?  Similarly with scan in/out

Reply:
I added  a comment about this.

The idea is that a scan in/scan out/scan enable may be top level ports (bterms) but also pins in dbInsts. Pins in cells (iterms) are used if the scan chain is part of a compressor/decompressor and the scan enable may be controlled by a TAP Controller in the case of boundary scan.

Reply:
Perhaps a quick picture showing what you have in mind would help me to understand.

Reply:
So, in this image, SI, SE, SO, and CLK would be bterms, because they are input ports at the top level of the module.

![Screenshot 2024-04-05 at 10-50-35 TestPrinciples pdf](https://github.com/The-OpenROAD-Project/OpenROAD/assets/1619948/a03303d0-f3fa-48a9-aed3-47eb5aed5872)

Basically:

```
module (SI, SE, CLK, SO);
   input SI, SE, CLK;
   output SO;
endmodule
````

An example of when SE would be an iterm, is for pipelined scan enabled:

![Screenshot 2024-04-05 at 10-52-49 Architecture for Pipelining of Scan Enable](https://github.com/The-OpenROAD-Project/OpenROAD/assets/1619948/c79c26cc-0b4e-4b41-85e5-727b2efc3efb)

Where the scan enabled signal comes from the output of the or gate.

For a case where CLK is coming from a gate output, we have clock gating:

![Screenshot 2024-04-05 at 10-51-57 TestPrinciples pdf](https://github.com/The-OpenROAD-Project/OpenROAD/assets/1619948/fd9e2242-9c92-492e-9817-4c6780e4626c)

GCK would be an iterm right?

And finally, for the case of compression, we will have SI and SO also as iterms (only showing the decompressor -> SI case here)

![Screenshot 2024-04-05 at 10-55-13 TestPrinciples pdf](https://github.com/The-OpenROAD-Project/OpenROAD/assets/1619948/157a2d86-5e5a-4722-ba13-005d79a025bf)

Again, this is from my understanding of the difference between bterms and iterms, I may be wrong here.

Thanks!

Reply:
I think I get it.  The key point is that the iterm isn't actually an iterm of any element in the dbScainChain but instead of some other control logic.